repology.1
.\"
.\" repology - Command line interface to Repology.org
.\"
.\" Copyright (c) 2024-2025 Brian Callahan <bcallah@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd January 5, 2025
.Dt REPOLOGY 1
.Os
.Sh NAME
.Nm repology
.Nd command line interface to Repology.org
.Sh SYNOPSIS
.Nm
.Op options
.Op Ar package ...
.Sh DESCRIPTION
.Nm
is a utility that retrieves package information from Repology.org.
.Pp
By default,
.Nm
selects an appropriate repository based on your operating system:
.Pp
.Dl Alpine Ar alpine_edge
.Dl Debian Ar debian_unstable
.Dl Fedora Ar fedora_rawhide
.Dl FreeBSD Ar freebsd
.Dl Gentoo Ar gentoo
.Dl macOS Ar homebrew
.Dl NetBSD Ar pkgsrc_current
.Dl OpenBSD Ar openbsd
.Dl OpenIndiana Ar openindiana
.Dl Ubuntu Ar ubuntu_ Ns Ar <releaseYear>_<releaseMonth>
.Dl Windows Ar chocolatey
.Pp
All other users will need to specify their repository with the
.Fl -repo
flag.
.Pp
.Nm
returns information in the following format:
.Pp
.Dl Port category/port name
.Dl Package version
.Dl Latest version
.Dl Maintainer email address
.Pp
Output is sorted by port category,
then port name.
.Pp
As per the API documentation,
.Nm
returns up to 200 items in a single API call.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl -repo Ar REPO
Override repository selection with
.Ar REPO .
.It Fl -begin Ar PROJECT
Get up to 200 projects starting with
.Ar PROJECT .
.It Fl -end Ar PROJECT
Get up to 200 projects ending with
.Ar PROJECT .
.It Fl -search Ar NAME
Project name substring to look for.
.It Fl -maintainer Ar EMAIL
Return projects maintained by specified
.Ar EMAIL .
.It Fl -category Ar NAME
Return projects in specified category.
.It Fl -inrepo Ar REPO
Return projects present in specified
.Ar REPO .
If this option is not issued,
.Ar REPO
will be set to the value of
.Fl -repo
or the auto-detected repository if the
.Fl -repo
option is also not issued.
.It Fl -notinrepo Ar REPO
Return projects absent in specified
.Ar REPO .
.It Fl -repos Ar NUMBER
Return projects present in specified
.Ar NUMBER
of repositories.
For all options that take a
.Ar NUMBER
argument,
exact values and open/closed ranges are allowed,
.Em e.g. ,
1, 5-, -3, 2-7.
.It Fl -families Ar NUMBER
Return projects present in specified
.Ar NUMBER
of repository families.
.It Fl -repos_newest Ar NUMBER
Return projects which are up to date in the specified
.Ar NUMBER
of repositories.
.It Fl -families_newest Ar NUMBER
Return projects which are up to date in the specified
.Ar NUMBER
of repository families.
.It Fl -newest
Return only packages marked newest.
.It Fl -outdated
Return only packages marked outdated.
.It Fl -problematic
Return only packages marked problematic.
.It Fl -sort_package
Sort output by package name instead of port category/port name.
Prepend output with package name.
.It Fl -all
Display package information for all known repositories in the following format:
.Pp
.Dl Repository
.Dl Package name
.Dl Package version
.Dl Latest version
.It Fl -version
Print version information and exit.
.It Fl -help
Print help information and exit.
.El
.Pp
Package version is colored based on its status.
The colors are as follows:
.Bl -tag -width Ds
.It Green
The package is up to date.
.It Red
The package is outdated.
.It Cyan
The package is either a development version, unique to this repo,
or a rolling release.
.It Yellow
The package is a legacy version.
.It Magenta
The package has no known versioning scheme.
.It Blue
The package version is incorrect, untrusted, or otherwise ignored.
.El
.Sh EXIT STATUS
The
.Nm
utility exits 0 on success, and >0 if an error occurs.
.Sh EXAMPLES
Get information on the Digital Mars D compiler from the auto-detected
repository:
.Pp
.Dl repology dmd
.Pp
Get information on the Digital Mars D compiler from all repositories:
.Pp
.Dl repology --all dmd
.Pp
Get 200 items from pkgsrc:
.Pp
.Dl repology --repo pkgsrc
.Pp
Get information about both Chrome and Firefox from
.Fx
ports:
.Pp
.Dl repology --repo freebsd chromium firefox
.Pp
Get up to 200 packages maintained by ports@openbsd.org,
starting with coreutils:
.Pp
.Dl repology --maintainer ports@openbsd.org --begin coreutils
.Sh SEE ALSO
The repology API documentation can be found at
.Lk https://repology.org/api/v1 .
.Sh AUTHORS
.Nm
was written by
.An Brian Callahan Aq Mt bcallah@openbsd.org .
.Sh CAVEATS
Not all repositories may contain all information.


NEWS
commit 9ea600063f317857342d7358ba43ae2aebe1b78c
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Mon Feb 10 20:59:01 2025 -0500

    Repology 1.9.0
    Add Gentoo

commit 87286b47496f1255fece927f45206b45ce381467
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Jan 5 20:35:31 2025 -0500

    Repology 1.8.0
    Support OpenIndiana
    Simplify linux autodetection

commit 41db3bccce0c634d04f3e27701424f978453a967
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Jan 5 18:59:17 2025 -0500

    Read in /etc/os-release directly

commit 3af0dbfaf3095331529697eb9b9c2a69985a340a
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 20:12:11 2024 -0500

    Update NEWS in preparation for release.

commit 993f2eed92d11129d6d0d85c2dc4da6a9c163792
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 20:11:09 2024 -0500

    Repology 1.7.3
    Switch Debian to unstable.
    Simplify Linux detection logic.

commit 60ab3e0b6b32c04099b4b090d9916969770c9918
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 19:50:34 2024 -0500

    Fix manual page to correct Debian and Fedora repos

commit 9e24cb113043eaa456e38ae6d112e28f427cc38c
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 19:40:32 2024 -0500

    Ubuntu supplies maintainer information.

commit b1ec1228f3ef07725026d100bdbf526c74bc364a
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 19:23:05 2024 -0500

    Default to the development branch of package repos

commit ecf7779e0d95030ae80771b362bd2ac3be18d36a
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 17:11:40 2024 -0500

    Update NEWS in preparation for release.

commit a40f7f3641b255d2e5186558ee7240eed1f2029d
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 17:11:00 2024 -0500

    Repology 1.7.0
    Support Alpine Linux.

commit 7cf43aa2aeca2beb77441f4d62d6c53919f7765f
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 16:26:09 2024 -0500

    Update NEWS in preparation for release.

commit 7c7a214d4507ad1ad7ce1d305774b595ab3e0c8b
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sun Dec 29 16:23:03 2024 -0500

    Repology 1.6.0
    Add autodetection for Fedora, Debian, and Ubuntu.

commit ce74d27918a08e6e8ef06fd8e92fafca8c5193b6
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Mon Dec 23 13:52:43 2024 -0500

    Update NEWS in preparation for release.

commit 2d721a79b44fdbff058820bdac57a2c1ae63885e
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Mon Dec 23 13:51:05 2024 -0500

    Repology 1.5.0
    Add autodetection for Windows, selecting cholocatey.
    Add a configure script to make packaging much easier.
    Improve documentation in manual page and README.md

commit c73c12c82d6412ecc3f529efdaf45c95f89ca12a
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Sat Dec 21 11:53:43 2024 -0500

    Repology 1.4.0
    Add support for macOS, defaulting to Homebrew

commit e332286e6798fa6cc0dd41c21b9c6dcb81493ba8
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Fri Jul 5 10:34:37 2024 -0400

    Repology 1.3.0
    Add an --all option that displays information about packages from
    all known repositories.

commit d4e08be2e83df4accdca8b2ffcb50edf23283843
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Thu Jul 4 18:41:43 2024 -0400

    Add a dub.json so repology can be published in code.dlang.org

commit 291dc269dd32a8f091af5a1c71ca0987c4c85bdc
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Thu Jul 4 17:33:46 2024 -0400

    Release version 1.2.0
    Fix some bugs that sometimes caused the most recent known version
    across all repos to be displayed incorrectly.
    Small improvements on overall display.

commit 39269e26c2af41c27fa157c606e16f2dbd8d0824
Merge: a89109c 19d3f58
Author: Brian Callahan <ibara@users.noreply.github.com>
Date:   Thu Jul 4 16:39:48 2024 -0400

    Merge pull request #1 from FeepingCreature/main
    
    Golf options parsing using compile-time introspection.

commit 19d3f5856a5e7533ccb547dfde54daf12db4836c
Author: Mathis Beer <Mathis.Beer@funkwerk-itk.com>
Date:   Wed Jul 3 11:11:59 2024 +0200

    Golf options parsing using compile-time introspection.

commit a89109ce86ec05fe2b0bf6ae1bae7350ff60a5af
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Tue Jul 2 19:02:04 2024 -0400

    repology-1.1.0
    Additional colors for status based on upstream.

commit 71f37be0f5c6014e99000c25d509713d3be8024e
Author: Brian Callahan <bcallah@openbsd.org>
Date:   Wed Jun 12 10:30:49 2024 -0400

    Initial commit of repology


repology.d
/+
 + Copyright (c) 2024-2025 Brian Callahan <bcallah@openbsd.org>
 +
 + Permission to use, copy, modify, and distribute this software for any
 + purpose with or without fee is hereby granted, provided that the above
 + copyright notice and this permission notice appear in all copies.
 +
 + THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 + WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 + MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 + ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 + WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 + ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 + OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 +/

import std.stdio;
import std.algorithm;
import std.getopt;
import std.json;
import std.meta;
import std.range;
import std.net.curl;

struct Options
{
    string repo,
           begin,
           end,
           search,
           maintainer,
           category,
           inrepo,
           notinrepo,
           repos,
           families,
           repos_newest,
           families_newest;

    bool newest,
         outdated,
         problematic,
         sort_package,
         all,
         vers;
}

int main(string[] args)
{
    Options options;
    JSONValue json;
    string[] pkgs;
    string uri = "https://repology.org/api/v1/project";
    string[string] queryParts = null;

    version (OpenBSD) options.repo = "openbsd";
    else version (FreeBSD) options.repo = "freebsd";
    else version (NetBSD) options.repo = "pkgsrc_current";
    else version (OSX) options.repo = "homebrew";
    else version (Windows) options.repo = "chocolatey";
    else version (Solaris) {
        import std.file;
        import std.string;
        string[] lines = splitLines(cast(string)read("/etc/release"));
        foreach (line; lines) {
            if (line.stripLeft.startsWith("OpenIndiana")) {
                options.repo = "openindiana";
                break;
            }
        }
        if (!options.repo)
            stderr.writeln("repology: specify your repo with the --repo flag");
    } else version (linux) {
        import std.file;
        import std.string;
        string distro, release;
        string[] lines = splitLines(cast(string)read("/etc/os-release"));
        foreach (line; lines) {
            if (line.startsWith("ID=")) {
                auto id = line.split("=");
                distro = id[1].idup;
            } else if (line.startsWith("VERSION_ID=")) {
                auto versionid = line.split("=");
                release = versionid[1].idup;
            }
        }
        switch (distro) {
        case "alpine":
            options.repo = distro ~ "_edge";
            break;
        case "debian":
            options.repo = distro ~ "_unstable";
            break;
        case "fedora":
            options.repo = distro ~ "_rawhide";
            break;
        case "gentoo":
            options.repo = distro;
            break;
        case "ubuntu":
            options.repo = distro ~ "_" ~
                release.strip("\"").replaceFirst(".", "_");
            break;
        default:
            stderr.writeln("repology: specify your repo with the --repo flag");
        }
    } else {
        stderr.writeln("repology: specify your repo with the --repo flag");
    }

    auto opts = getopt(
        args,
        "repo", "string", &options.repo,
        "begin", "string", &options.begin,
        "end", "string", &options.end,
        "search", "string", &options.search,
        "maintainer", "string", &options.maintainer,
        "category", "string", &options.category,
        "inrepo", "string", &options.inrepo,
        "notinrepo", "string", &options.notinrepo,
        "repos", "string", &options.repos,
        "families", "string", &options.families,
        "repos_newest", "string", &options.repos_newest,
        "families_newest", "string", &options.families_newest,
        "newest", "boolean", &options.newest,
        "outdated", "boolean", &options.outdated,
        "problematic", "boolean", &options.problematic,
        "sort_package", "boolean", &options.sort_package,
        "all", "boolean", &options.all,
        "version", "Print version information.", &options.vers
    );

    if (opts.helpWanted) {
        defaultGetoptPrinter("usage: repology [options] [package ...]",
            opts.options);
        return 1;
    }

    if (options.vers) {
        writeln("1.9.0 (10 Feb 2025)");
        return 1;
    }

    switch (options.repo) {
    case "alpine":
        options.repo = "alpine_edge";
        break;
    case "debian":
        options.repo = "debian_unstable";
        break;
    case "fedora":
        options.repo = "fedora_rawhide";
        break;
    case "pkgsrc":
        options.repo = "pkgsrc_current";
        break;
    default:
        {}
    }

    if (!options.all) {
        if (!options.inrepo)
            options.inrepo = options.repo;
    }

    enum queryOptions = [__traits(allMembers, Options)]
        .filter!(a => !["repo", "begin", "end", "repos", "sort_package"].canFind(a));
    static foreach (member; queryOptions) {
        if (mixin(`options.` ~ member)) {
            alias T = typeof(mixin(`options.` ~ member));
            static if (is(T == bool))
                queryParts[member] = "1";
            else
                queryParts[member] = mixin(`options.` ~ member);
        }
    }

    string query = "?" ~ queryParts.byKeyValue.map!(a => a.key ~ "=" ~ a.value).join("&");
    if (args.length == 1) {
        uri ~= "s/";
        if (options.begin)
            uri ~= options.begin ~ "/";
        else if (options.end)
            uri ~= ".." ~ options.end ~ "/";
        auto res = get(uri ~ query);
        json = parseJSON(res);
        foreach (obj; json.object)
            pkgs ~= process(obj, options);
    } else {
        foreach (arg; args[1 .. $]) {
            auto res = get(uri ~ "/" ~ arg ~ query);
            json = parseJSON(res);
            pkgs ~= process(json, options);
        }
    }
    pkgs.sort;
    foreach (pkg; pkgs) {
        if (!pkg.empty)
            writeln(pkg);
    }

    return 0;
}

string[] process(JSONValue json, Options options)
{
    JSONValue[] j;
    string[] info;
    string latest;
    bool hasmaintainer, havelatest;

    switch (options.repo) {
    case "alpine_3_20":
    case "alpine_3_21":
    case "alpine_edge":
    case "debian_10":
    case "debian_11":
    case "debian_12":
    case "debian_13":
    case "debian_experimental":
    case "debian_unstable":
    case "freebsd":
    case "gentoo":
    case "openbsd":
    case "pkgsrc_current":
    case "ubuntu_24_04":
    case "ubuntu_24_04_backports":
    case "ubuntu_24_10":
    case "ubuntu_24_10_backports":
    case "ubuntu_25_04":
    case "ubuntu_25_04_proposed":
        hasmaintainer = true;
        break;
    default:
        hasmaintainer = false;
    }

    foreach (obj; json.array) {
        switch (obj["status"].str) {
        case "newest":
            latest = obj["version"].str;
            havelatest = true;
            break;
        case "devel":
        case "unique":
            if (!havelatest) {
                latest = obj["version"].str;
                havelatest = true;
            }
            break;
        case "noscheme":
            if (!havelatest) {
                latest = "noscheme";
                havelatest = true;
            }
            break;
        case "rolling":
            if (!havelatest) {
                latest = "rolling";
                havelatest = true;
            }
            break;
        default:
            break;
        }
        if (options.all) {
            j ~= obj;
        } else {
            if (obj["repo"].str == options.repo)
                j ~= obj;
        }
    }

    foreach (obj; j) {
        string tmp;
        if (options.all) {
            tmp ~= obj["repo"].str ~ " ";
            tmp ~= obj["visiblename"].str ~ " ";
        }
        if (!options.all) {
            if (options.sort_package)
                tmp ~= obj["binname"].str ~ " ";
            if (options.repo == "chocolatey")
                tmp ~= obj["binname"].str ~ " ";
            else
                tmp ~= obj["srcname"].str ~ " ";
        }
        string color;
        switch (obj["status"].str) {
        case "newest":
            color = "32";
            break;
        case "outdated":
            color = "31";
            break;
        case "devel":
        case "rolling":
        case "unique":
            color = "36";
            break;
        case "legacy":
            color = "33";
            break;
        case "noscheme":
            color = "35";
            break;
        case "incorrect":
        case "untrusted":
        case "ignored":
            color = "34";
            break;
        default:
            color = "32";
        }
        string ver = "\033[" ~ color ~ "m" ~ obj["version"].str ~ "\033[0m";
        if (!options.all && hasmaintainer) {
            string maintainer = "";
            auto maintainers = obj["maintainers"].array;
            foreach (i; 0 .. maintainers.length) {
                maintainer ~= maintainers[i].str;
                if (i < maintainers.length - 1)
                    maintainer ~= " ";
            }
            tmp ~= ver ~ " " ~ latest ~ " " ~ maintainer;
        } else {
            tmp ~= ver ~ " " ~ latest;
        }
        info ~= tmp;
    }

    return info;
}
